SerialPrograms/Source/PokemonLA/Options/PokemonLA_BattlePokemonActionTable.cpp

```diff
@@ -251,6 +251,109 @@ MoveStyle OneMoveBattlePokemonActionTable::get_style(size_t pokemon){
+            const auto iter = MoveStyle_MAP.find(value);
+            if (iter != MoveStyle_MAP.end()) {
+                m_style = iter->second;
+            }
+        }
+    }
+    json_get_int(m_attemps, obj, "Attempts", 0);
+}
+
+QJsonValue MoveGrinderActionRow::to_json() const
+{
+    QJsonObject obj;
+    obj.insert("PokemonIndex", m_pokemon_index);
+    obj.insert("MoveIndex", m_move_index);
+    obj.insert("Style", MoveStyle_NAMES[(size_t)m_style]);
+    obj.insert("Attempts", m_attemps);
+    return obj;
+}
+
+std::unique_ptr<EditableTableRow> MoveGrinderActionRow::clone() const
+{
+    return std::unique_ptr<EditableTableRow>(new MoveGrinderActionRow(*this));
+}
+
+std::vector<QWidget*> MoveGrinderActionRow::make_widgets(QWidget& parent)
+{
+    std::vector<QWidget*> widgets;
+    widgets.emplace_back(make_conststring_table_cell(parent, m_pokemon_index));
+    widgets.emplace_back(make_conststring_table_cell(parent, m_move_index));
+    widgets.emplace_back(make_enum_table_cell(parent, MoveStyle_MAP.size(), MoveStyle_NAMES, m_style));
+    widgets.emplace_back(make_integer_table_cell(parent, m_attemps));
+    return widgets;
+}
+
+QStringList MoveGrinderActionTableFactory::make_header() const
+{
+    QStringList list;
+    list << "Pokemon index" << "Move index" << "Move Style" << "Move Attempts";
+    return list;
+}
+
+std::unique_ptr<EditableTableRow> MoveGrinderActionTableFactory::make_row() const
+{
+    return std::unique_ptr<EditableTableRow>(new MoveGrinderActionRow("UNUSED", 0));
+}
+
+MoveGrinderActionTable::MoveGrinderActionTable()
+    : m_table("For every move you want to perform, input the style and the number of attmps you want to achieve.", m_factory, make_defaults())
+{}
+
+std::optional<MoveStyle> MoveGrinderActionTable::get_style(size_t pokemon, size_t move) const
+{
+    const MoveGrinderActionRow& action = static_cast<const MoveGrinderActionRow&>(m_table[pokemon * 4 + move]);
+    if (action.m_attemps == 0)
+    {
+        return {};
+    }
+    return action.m_style;
+}
+
+void MoveGrinderActionTable::use_move(size_t pokemon, size_t move)
+{
+    MoveGrinderActionRow& action = static_cast<MoveGrinderActionRow&>(m_table[pokemon * 4 + move]);
+    action.m_attemps--;
+}
+
+void MoveGrinderActionTable::load_json(const QJsonValue& json) {
+    m_table.load_json(json);
+}
+
+QJsonValue MoveGrinderActionTable::to_json() const {
+    return m_table.to_json();
+}
+
+ConfigWidget* MoveGrinderActionTable::make_ui(QWidget& parent) {
+    return m_table.make_ui(parent);
+}
+
+std::vector<std::unique_ptr<EditableTableRow>> MoveGrinderActionTable::make_defaults() const {
+    std::vector<std::unique_ptr<EditableTableRow>> ret;
+    for (int i = 0; i < 5; ++i)
+    {
+        for (int j = 0; j < 4; ++j)
+        {
+            ret.emplace_back(std::make_unique<MoveGrinderActionRow>(STRING_POKEMON + " " + QString::number(i), "Move " + QString::number(j)));
+        }
+    }
+    return ret;

```

Comment:
Is there a way to "freeze" the table, meaning disable the add/remove, etc ? If not, maybe I should create a new structure for this and the issue above

Reply:
Other issue is fixed and hopefully the current display will make it clear that the user isn't supposed to add/remove lines but well